components/base/sensorcontrolled/sensorcontrolled.go

```diff
@@ -279,3 +280,47 @@ func (sb *sensorBase) Close(ctx context.Context) error {
 	sb.activeBackgroundWorkers.Wait()
 	return nil
 }
+
+// determineHeadingFunc determines which movement sensor endpoint should be used for control.
+// The priority is Orientation -> Heading -> No heading control.
+func (sb *sensorBase) determineHeadingFunc(ctx context.Context) {
+	switch {
+	case sb.orientation != nil:
+		sb.logger.CInfof(ctx, "using orientation for base %v's MoveStraight heading control and Spin",
+			sb.Name().ShortName())
+		sb.headingFunc = func(ctx context.Context) (float64, error) {
+			orient, err := sb.orientation.Orientation(ctx, nil)
+			if err != nil {
+				return 0, err
+			}
+			// this returns (-180-> 180)
+			yaw := rdkutils.RadToDeg(orient.EulerAngles().Yaw)
+
+			return yaw, nil
+		}
+	case sb.compassHeading != nil:
+		sb.logger.CInfof(ctx, "using compass heading for base %v's MoveStraight heading control and Spin",
+			sb.Name().ShortName())
+		sb.headingFunc = func(ctx context.Context) (float64, error) {
+			compassHeading, err := sb.compassHeading.CompassHeading(ctx, nil)
+			if err != nil {
+				return 0, err
+			}
+			// flip compass heading to be CCW/Z up
+			compassHeading = 360 - compassHeading
+
+			// make the compass heading (-180->180)
+			if compassHeading > 180 {
+				compassHeading -= 360
+			}
```

Comment:
We write this so much that we should have a math helper function in utils. Can stay as is for now.